feat(options): add support for `--mode` flag (`options.mode`)

cilliemalan opened the pull request:
<!--
  HOLY CRAP a Pull Request. We ❤️ those!

  If you remove or skip this template, you'll make the 🐼 sad and the mighty god
  of Github will appear and pile-drive the close button from a great height
  while making animal noises.

  Please place an x (no spaces!) in all [ ] that apply
-->

This PR contains a:

- [ ] **bugfix**
- [x] new **feature**
- [ ] **code refactor**
- [ ] **test update** <!-- if bug or feature is checked, this should be too -->
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case
This PR resolves #225.

There is currently no way to specify `mode` when run against a project utilizing [Webpack 4 zero-configuration](https://webpack.js.org/guides/getting-started/#using-a-configuration). If webpack-serve is run on a project that doesn't have any configuration, it will always run in production mode and produce a warning.

This PR adds support for a `--mode` command line argument or `mode` option that allows one to specify the mode in which Webpack will compile (development, production, or none).
<!--
  Please explain the motivation or use-case for your change.
  What existing problem does the PR solve?
  If this PR addresses an issue, please link to the issue.
-->

### Breaking Changes

<!--
  If this PR introduces a breaking change, please describe the impact and a
  migration path for existing applications.
-->

### Additional Info
If `--mode` is specified and a configuration file *is* present, the value specified will override what is configured. This is the same behaiour as `webpack-cli`.

cilliemalan committed:
feat: add support for `--mode` option and argument

Adding `--mode` makes it possible to set mode when using zero-config with Webpack 4.

resolves #225

shellscape commented on README.md:
```diff
@@ -67,6 +67,8 @@ Options
   --hmr            Specify whether or not the client should apply Hot Module Replacement patches (default: true)
   --reload         Specify whether or not the middleware should reload the page for build errors (default: true)
+  --mode           Enable production optimizations or development hints.
+                   Choices: development, production, none
```
Please change this to: `Modes: `

shellscape commented on lib/config.js:
```diff
@@ -119,6 +120,11 @@ module.exports = {
 
       found = [].concat(found).map((conf) => prepare(conf));
 
+      if (mode) {
```
not the right spot to do this. the flag value should be applied to the `result` in the `apply()` method of `lib/flags.js`. it's impossible to know which config in a multi-config that a user intends to apply a flag to, so it should apply to all configs.

shellscape commented on schemas/flags.js:
```diff
@@ -14,7 +14,8 @@ module.exports = {
   'dev-ware': {
-    desc: 'Set options for webpack-dev-middleware. e.g. --dev-ware.publicPath /',
+    desc:
+      'Set options for webpack-dev-middleware. e.g. --dev-ware.publicPath /',
```
please keep an eye on your editor making unnecessary style changes

cilliemalan replied:
I'll fix it but note that `lint-staged` is doing this.

shellscape commented on schemas/flags.js:
```diff
@@ -70,6 +71,11 @@ Use {dim --no-hot-client} to disable webpack-hot-client`,
   },
+  mode: {
+    desc: chalk`Enable production optimizations or development hints.
+{dim Choices: development, production, none}`,
```
`Modes: ` please

shellscape commented on test/config.test.js:
```diff
@@ -122,4 +122,20 @@ describe('config', () => {
+
+  test('load: mode overrides config', () => {
+    const argv = { mode: 'production' };
+    const config = './test/fixtures/webpack-4-defaults/webpack.config.js';
```
this is a good test, but we also need a test with the multi-config to make sure the mode was applied correctly to an array of configs

shellscape reviewed (changes requested):
great start, see comments

cilliemalan commented:
@shellscape I've made most of the changes you requested.

I've implemented the mode override in `apply` in `lib/options`. I don't quite see how it'll work in `lib/flags` as the config is not available there. Let me know if this is okay?

I've added tests for `options`, checking that the mode is set in single config and multiple config. I've also added a test fixture and two tests for Webpack 4 no-config (one normal one, and one checking that mode is set on the compiler when passed as an argument).

Let me know if this is okay!

cilliemalan commented:
Fixed the test.

G-Rath reviewed (approved):
@shellscape Looks decent.

As cilliemalan said, I'm not sure how you'd implement the mode override in `lib/options`, unless maybe you also change `lib/compiler` to actually look for that property in the `options` argument...?

That's the only part I think you need to double check - the rest looks solid, and seems to properly implement all the changes you requested in your last review.

alexander-akait commented:
DEPRECATED. Please use `webpack-dev-server` (he is in support and update mode again). Feel free to create a issue if some features are not implemented in `webpack-dev-server`. 

Why deprecated `webpack-serve` ?
- The author stopped developing the package.
- Two development servers are misleading for developers.
- Hard to maintain two packages with same purpose.
- Some `dependencies` are not available to fixes/implement new features/etc (https://www.npmjs.com/package/webpack-hot-client and https://github.com/shellscape/koa-static) and there's nothing we can do here.
- `webpack-dev-server` is stable.

Thanks for using `webpack`! We apologize for the inconvenience. In the future, we will avoid such situations.
